src/shared/port-provider.js

```diff
@@ -86,67 +80,114 @@ export class PortProvider {
     //   neutered.
     // - Messages are queued until the other port is ready to listen (`port.start()`)
 
-    // Create the `host-sidebar` channel immediately, while other channels are
+    // Create the `sidebar-host` channel immediately, while other channels are
     // created on demand
-    this._hostSidebarChannel = new MessageChannel();
+    this._sidebarHostChannel = new MessageChannel();
 
     this._listeners = new ListenerCollection();
+
+    /** @type {Array<Message & {allowedOrigin: string}>} */
+    this._allowedMessages = [
+      {
+        allowedOrigin: '*',
+        authority: 'hypothesis',
+        frame1: 'guest',
+        frame2: 'host',
+        type: 'request',
+      },
+      {
+        allowedOrigin: '*',
+        authority: 'hypothesis',
+        frame1: 'guest',
+        frame2: 'sidebar',
+        type: 'request',
+      },
+      {
+        allowedOrigin: this._hypothesisAppsOrigin,
+        authority: 'hypothesis',
+        frame1: 'sidebar',
+        frame2: 'host',
+        type: 'request',
+      },
+      {
+        allowedOrigin: this._hypothesisAppsOrigin,
+        authority: 'hypothesis',
+        frame1: 'notebook',
+        frame2: 'sidebar',
+        type: 'request',
+      },
+    ];
   }
 
   /**
-   * Checks the `postMessage` origin and message.
+   * Check that source is of type Window.
    *
-   * @param {MessageEvent} event
-   * @param {Message} allowedMessage - the MessageEvent's data must match this
-   *   object to grant the port.
-   * @param {string} allowedOrigin - the MessageEvent's origin must match this
-   *   value to grant the port. If '*' allow all origins.
+   * @param {MessageEventSource|null} source
+   * @return {source is Window}
    */
-  _isValidRequest(event, allowedMessage, allowedOrigin) {
-    const { data, origin, source } = event;
-    if (allowedOrigin !== '*' && origin !== allowedOrigin) {
-      return false;
-    }
-
+  _isSourceWindow(source) {
     if (
       // `source` can be of type Window | MessagePort | ServiceWorker.
       // The simple check `source instanceof Window`` doesn't work here.
       // Alternatively, `source` could be casted `/** @type{Window} */ (source)`
-      !source ||
+      source === null ||
       source instanceof MessagePort ||
       source instanceof ServiceWorker
```

Comment:
This currently covers all the kinds of `MessageEventSource` that exist. In future new ones might be added. Since we check the content of the message as well as the source before processing it, I don't think that should be a problem though.